File: Main/mess2.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <unordered_map>
#include <cmath>
#include <algorithm>
#include <omp.h>
#include <chrono>
#include <set>
#include <string>
#include <random>
#include <queue>
#include <regex>

class Graph {
private:
    std::unordered_map<int, std::vector<int>> adjacencyList;
    std::unordered_map<int, double> nodeDegrees;
    double totalEdges = 0;

public:
    void addEdge(int from, int to) {
        adjacencyList[from].push_back(to);
        adjacencyList[to].push_back(from);
        nodeDegrees[from]++;
        nodeDegrees[to]++;
        totalEdges++;
    }

    std::vector<int> getNeighbors(int node) {
        return adjacencyList[node];
    }

    double getDegree(int node) {
        return nodeDegrees[node];
    }

    double getTotalEdges() {
        return totalEdges;
    }

    std::vector<int> getNodes() {
        std::vector<int> nodes;
        for (const auto& pair : adjacencyList) {
            nodes.push_back(pair.first);
        }
        return nodes;
    }
};

class Message {
public:
    enum class State { CREATED, SHARED, VIRAL, FLAGGED };
    
    Message(int id, const std::string& content, int sourceNode)
        : id(id), content(content), sourceNode(sourceNode), state(State::CREATED), shareCount(0) {}
    
    int getId() const { return id; }
    const std::string& getContent() const { return content; }
    int getSourceNode() const { return sourceNode; }
    State getState() const { return state; }
    int getShareCount() const { return shareCount; }
    
    void incrementShareCount() { 
        shareCount++; 
        updateState();
    }
    
    void flagAsMisinformation() { state = State::FLAGGED; }

private:
    int id;
    std::string content;
    int sourceNode;
    State state;
    int shareCount;
    
    void updateState() {
        if (shareCount > 100) state = State::VIRAL;
        else if (shareCount > 10) state = State::SHARED;
    }
};

class LouvainCommunityDetection {
private:
    Graph graph;
    std::vector<int> communities;
    double modularity;
    std::vector<Message> messages;
    std::mt19937 rng;

    double calculateModularity() {
        std::vector<double> communityInternalEdges(communities.size(), 0);
        std::vector<double> communityTotalEdges(communities.size(), 0);
        double m = graph.getTotalEdges();
        double Q = 0;

        #pragma omp parallel for
        for (int i = 0; i < graph.getNodes().size(); i++) {
            int node = graph.getNodes()[i];
            int comm = communities[node];
            double nodeDegree = graph.getDegree(node);

            #pragma omp atomic
            communityTotalEdges[comm] += nodeDegree;

            for (int neighbor : graph.getNeighbors(node)) {
                if (communities[neighbor] == comm) {
                    #pragma omp atomic
                    communityInternalEdges[comm] += 1;
                }
            }
        }

        for (int i = 0; i < communities.size(); i++) {
            if (communityTotalEdges[i] > 0) {
                Q += (communityInternalEdges[i] / (2 * m)) - std::pow(communityTotalEdges[i] / (2 * m), 2);
            }
        }

        return Q;
    }

    void moveNode(int node) {
        int currentCommunity = communities[node];
        std::unordered_map<int, double> communityGains;

        for (int neighbor : graph.getNeighbors(node)) {
            int neighborCommunity = communities[neighbor];
            if (communityGains.find(neighborCommunity) == communityGains.end()) {
                communityGains[neighborCommunity] = 0;
            }
            communityGains[neighborCommunity] += 1;
        }

        int bestCommunity = currentCommunity;
        double bestGain = 0;

        for (const auto& pair : communityGains) {
            int community = pair.first;
            double gain = pair.second - (graph.getDegree(node) * communityGains[community] / (2 * graph.getTotalEdges()));
            if (gain > bestGain) {
                bestGain = gain;
                bestCommunity = community;
            }
        }

        if (bestCommunity != currentCommunity) {
            communities[node] = bestCommunity;
        }
    }

public:
    LouvainCommunityDetection() : rng(std::random_device{}()) {}

    void loadGraph(const std::string& filename) {
        std::ifstream file(filename);
        std::string line;

        while (std::getline(file, line)) {
            if (line[0] == '#') continue;
            std::istringstream iss(line);
            int from, to;
            if (iss >> from >> to) {
                graph.addEdge(from, to);
            }
        }

        communities.resize(graph.getNodes().size());
        for (int i = 0; i < graph.getNodes().size(); i++) {
            communities[i] = i;
        }
    }

    int detectCommunities(bool useParallel = false) {
        bool improvement = true;
        while (improvement) {
            improvement = false;
            if (useParallel) {
                #pragma omp parallel for schedule(dynamic)
                for (int i = 0; i < graph.getNodes().size(); i++) {
                    int node = graph.getNodes()[i];
                    int oldCommunity = communities[node];
                    moveNode(node);
                    if (communities[node] != oldCommunity) {
                        #pragma omp atomic write
                        improvement = true;
                    }
                }
            } else {
                for (int node : graph.getNodes()) {
                    int oldCommunity = communities[node];
                    moveNode(node);
                    if (communities[node] != oldCommunity) {
                        improvement = true;
                    }
                }
            }
        }

        modularity = calculateModularity();

        std::unordered_map<int, int> uniqueCommunities;
        for (int community : communities) {
            uniqueCommunities[community] = 1;
        }

        return uniqueCommunities.size();
    }

    double getModularity() {
        return modularity;
    }

    struct NodeInfo {
        int community;
        std::set<int> connectedCommunities;
        std::vector<int> directlyConnectedNodes;
        std::set<int> allConnectedNodes;
    };

    NodeInfo getNodeInfo(int targetNode) {
        NodeInfo info;
        info.community = communities[targetNode];

        std::set<int> visited;
        std::vector<int> queue;

        queue.push_back(targetNode);
        visited.insert(targetNode);

        while (!queue.empty()) {
            int currentNode = queue.front();
            queue.erase(queue.begin());

            for (int neighbor : graph.getNeighbors(currentNode)) {
                if (currentNode == targetNode) {
                    info.directlyConnectedNodes.push_back(neighbor);
                }

                info.connectedCommunities.insert(communities[neighbor]);
                info.allConnectedNodes.insert(neighbor);

                if (visited.find(neighbor) == visited.end()) {
                    visited.insert(neighbor);
                    queue.push_back(neighbor);
                }
            }
        }

        return info;
    }

    void initiateMessage(int sourceNode, const std::string& content) {
        messages.emplace_back(messages.size(), content, sourceNode);
        propagateMessage(messages.back());
    }
    
    std::set<int> propagateMessage(const Message& message, int startNode = -1) {
        std::queue<int> nodesToProcess;
        std::set<int> affectedNodes;
        
        if (startNode == -1) {
            startNode = message.getSourceNode();
        }
        
        nodesToProcess.push(startNode);
        affectedNodes.insert(startNode);
        
        while (!nodesToProcess.empty()) {
            int currentNode = nodesToProcess.front();
            nodesToProcess.pop();
            
            for (int neighbor : graph.getNeighbors(currentNode)) {
                if (affectedNodes.count(neighbor) == 0 && shouldShareMessage()) {
                    messages[message.getId()].incrementShareCount();
                    nodesToProcess.push(neighbor);
                    affectedNodes.insert(neighbor);
                }
            }
        }
        
        return affectedNodes;
    }
    
    bool shouldShareMessage() {
        std::uniform_real_distribution<> dis(0.0, 1.0);
        return dis(rng) < 0.3; // 30% chance of sharing
    }
    
    bool isMisinformation(const Message& message, double spreadPercentage) {
        static const std::regex misinfoPattern("\\b(fake|hoax|conspiracy)\\b");
        return std::regex_search(message.getContent(), misinfoPattern) || spreadPercentage > 0.1; // 10% threshold
    }
    
    void analyzeMessageImpact(int targetNode, const std::string& messageContent = "") {
        std::string content = messageContent.empty() ? "Sample message from target node" : messageContent;
        Message targetMessage(messages.size(), content, targetNode);
        messages.push_back(targetMessage);
        
        std::set<int> affectedNodes = propagateMessage(targetMessage, targetNode);
        double spreadPercentage = static_cast<double>(affectedNodes.size()) / graph.getNodes().size();
        
        std::cout << "Message from target node " << targetNode << ":" << std::endl;
        std::cout << "Content: " << content << std::endl;
        std::cout << "Affected nodes: " << affectedNodes.size() << std::endl;
        std::cout << "Spread percentage: " << spreadPercentage * 100 << "%" << std::endl;
        
        if (isMisinformation(targetMessage, spreadPercentage)) {
            std::cout << "This message has been flagged as potential misinformation." << std::endl;
        } else {
            std::cout << "This message has not been flagged as misinformation." << std::endl;
        }
        
        // Check if the target node has any flagged misinformation messages
        bool hasMisinfoMessage = false;
        for (const auto& message : messages) {
            if (message.getSourceNode() == targetNode && message.getState() == Message::State::FLAGGED) {
                hasMisinfoMessage = true;
                break;
            }
        }
        
        std::cout << "Target node " << (hasMisinfoMessage ? "has" : "does not have") << " flagged misinformation messages." << std::endl;
    }
};

int main() {
    LouvainCommunityDetection lcd;
    lcd.loadGraph("sample_graph1500.txt");

    // Parallel execution
    auto start_parallel = std::chrono::high_resolution_clock::now();
    int numCommunitiesParallel = lcd.detectCommunities(true);
    auto end_parallel = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> parallel_time = end_parallel - start_parallel;

    std::cout << "Number of communities detected: " << numCommunitiesParallel << std::endl;
    std::cout << "Parallel execution time: " << parallel_time.count() << " seconds" << std::endl;

    // Initiate some sample messages
    lcd.initiateMessage(1, "This is a normal message.");
    lcd.initiateMessage(10, "FAKE: Earth is flat! Share this conspiracy theory!");
    lcd.initiateMessage(100, "COVID-19 vaccine contains microchips. This is a hoax!");

    // Get target node from user
    int targetNode;
    std::cout << "\nEnter a target node: ";
    std::cin >> targetNode;

    // Get message content from user (optional)
    std::string messageContent;
    std::cout << "Enter a message for the target node (press Enter for default): ";
    std::cin.ignore();
    std::getline(std::cin, messageContent);

    // Analyze message impact from target node
    lcd.analyzeMessageImpact(targetNode, messageContent);

    // Get and display node information
    LouvainCommunityDetection::NodeInfo nodeInfo = lcd.getNodeInfo(targetNode);

    std::cout << "\nTarget Node: " << targetNode << std::endl;
    std::cout << "Community: " << nodeInfo.community << std::endl;
    std::cout << "Number of connected communities: " << nodeInfo.connectedCommunities.size() << std::endl;
    std::cout << "Number of directly connected nodes: " << nodeInfo.directlyConnectedNodes.size() << std::endl;
    std::cout << "Total number of connected nodes from all communities: " << nodeInfo.allConnectedNodes.size() << std::endl;

    std::cout << "Directly connected nodes: ";
    for (int node : nodeInfo.directlyConnectedNodes) {
        std::cout << node << " ";
    }
    std::cout << std::endl;

    return 0;
}

File: Main/mess2.py
import networkx as nx
from collections import defaultdict
import random
import time
import re

class Graph:
    def __init__(self):
        self.adjacency_list = defaultdict(list)
        self.node_degrees = defaultdict(float)
        self.total_edges = 0

    def add_edge(self, from_node, to_node):
        self.adjacency_list[from_node].append(to_node)
        self.adjacency_list[to_node].append(from_node)
        self.node_degrees[from_node] += 1
        self.node_degrees[to_node] += 1
        self.total_edges += 1

    def get_neighbors(self, node):
        return self.adjacency_list[node]

    def get_degree(self, node):
        return self.node_degrees[node]

    def get_total_edges(self):
        return self.total_edges

    def get_nodes(self):
        return list(self.adjacency_list.keys())

class Message:
    class State:
        CREATED = 0
        SHARED = 1
        VIRAL = 2
        FLAGGED = 3

    def __init__(self, id, content, source_node):
        self.id = id
        self.content = content
        self.source_node = source_node
        self.state = self.State.CREATED
        self.share_count = 0

    def get_id(self):
        return self.id

    def get_content(self):
        return self.content

    def get_source_node(self):
        return self.source_node

    def get_state(self):
        return self.state

    def get_share_count(self):
        return self.share_count

    def increment_share_count(self):
        self.share_count += 1
        self.update_state()

    def flag_as_misinformation(self):
        self.state = self.State.FLAGGED

    def update_state(self):
        if self.share_count > 100:
            self.state = self.State.VIRAL
        elif self.share_count > 10:
            self.state = self.State.SHARED

class LouvainCommunityDetection:
    def __init__(self):
        self.graph = Graph()
        self.communities = []
        self.modularity = 0
        self.messages = []
        self.rng = random.Random()

    def calculate_modularity(self):
        community_internal_edges = [0] * len(self.communities)
        community_total_edges = [0] * len(self.communities)
        m = self.graph.get_total_edges()
        Q = 0

        for node in self.graph.get_nodes():
            comm = self.communities[node]
            node_degree = self.graph.get_degree(node)
            community_total_edges[comm] += node_degree

            for neighbor in self.graph.get_neighbors(node):
                if self.communities[neighbor] == comm:
                    community_internal_edges[comm] += 1

        for i in range(len(self.communities)):
            if community_total_edges[i] > 0:
                Q += (community_internal_edges[i] / (2 * m)) - (community_total_edges[i] / (2 * m)) ** 2

        return Q

    def move_node(self, node):
        current_community = self.communities[node]
        community_gains = defaultdict(float)

        for neighbor in self.graph.get_neighbors(node):
            neighbor_community = self.communities[neighbor]
            community_gains[neighbor_community] += 1

        best_community = current_community
        best_gain = 0

        for community, gain in community_gains.items():
            gain -= (self.graph.get_degree(node) * community_gains[community] / (2 * self.graph.get_total_edges()))
            if gain > best_gain:
                best_gain = gain
                best_community = community

        if best_community != current_community:
            self.communities[node] = best_community

    def load_graph(self, filename):
        with open(filename, 'r') as file:
            for line in file:
                if line[0] == '#':
                    continue
                from_node, to_node = map(int, line.split())
                self.graph.add_edge(from_node, to_node)

        self.communities = list(range(len(self.graph.get_nodes())))

    def detect_communities(self, use_parallel=False):
        improvement = True
        while improvement:
            improvement = False
            for node in self.graph.get_nodes():
                old_community = self.communities[node]
                self.move_node(node)
                if self.communities[node] != old_community:
                    improvement = True

        self.modularity = self.calculate_modularity()

        unique_communities = set(self.communities)
        return len(unique_communities)

    def get_modularity(self):
        return self.modularity

    class NodeInfo:
        def __init__(self):
            self.community = 0
            self.connected_communities = set()
            self.directly_connected_nodes = []
            self.all_connected_nodes = set()

    def get_node_info(self, target_node):
        info = self.NodeInfo()
        info.community = self.communities[target_node]

        visited = set()
        queue = [target_node]
        visited.add(target_node)

        while queue:
            current_node = queue.pop(0)
            for neighbor in self.graph.get_neighbors(current_node):
                if current_node == target_node:
                    info.directly_connected_nodes.append(neighbor)

                info.connected_communities.add(self.communities[neighbor])
                info.all_connected_nodes.add(neighbor)

                if neighbor not in visited:
                    visited.add(neighbor)
                    queue.append(neighbor)

        return info

    def initiate_message(self, source_node, content):
        message = Message(len(self.messages), content, source_node)
        self.messages.append(message)
        self.propagate_message(message)

    def propagate_message(self, message, start_node=-1):
        if start_node == -1:
            start_node = message.get_source_node()

        nodes_to_process = [start_node]
        affected_nodes = set([start_node])

        while nodes_to_process:
            current_node = nodes_to_process.pop(0)
            for neighbor in self.graph.get_neighbors(current_node):
                if neighbor not in affected_nodes and self.should_share_message():
                    self.messages[message.get_id()].increment_share_count()
                    nodes_to_process.append(neighbor)
                    affected_nodes.add(neighbor)

        return affected_nodes

    def should_share_message(self):
        return self.rng.random() < 0.3

    def is_misinformation(self, message, spread_percentage):
        misinfo_pattern = re.compile(r'\b(fake|hoax|conspiracy)\b')
        return bool(misinfo_pattern.search(message.get_content())) or spread_percentage > 0.1

    def analyze_message_impact(self, target_node, message_content=""):
        content = message_content if message_content else "Sample message from target node"
        target_message = Message(len(self.messages), content, target_node)
        self.messages.append(target_message)

        affected_nodes = self.propagate_message(target_message, target_node)
        spread_percentage = len(affected_nodes) / len(self.graph.get_nodes())

        print(f"Message from target node {target_node}:")
        print(f"Content: {content}")
        print(f"Affected nodes: {len(affected_nodes)}")
        print(f"Spread percentage: {spread_percentage * 100}%")

        if self.is_misinformation(target_message, spread_percentage):
            print("This message has been flagged as potential misinformation.")
        else:
            print("This message has not been flagged as misinformation.")

        has_misinfo_message = any(
            message.get_source_node() == target_node and message.get_state() == Message.State.FLAGGED
            for message in self.messages
        )

        print(f"Target node {'has' if has_misinfo_message else 'does not have'} flagged misinformation messages.")

def main():
    lcd = LouvainCommunityDetection()
    lcd.load_graph("sample_graph1500.txt")

    start_time = time.time()
    num_communities = lcd.detect_communities(True)
    end_time = time.time()
    parallel_time = end_time - start_time

    print(f"Number of communities detected: {num_communities}")
    print(f"Parallel execution time: {parallel_time} seconds")

    lcd.initiate_message(1, "This is a normal message.")
    lcd.initiate_message(10, "FAKE: Earth is flat! Share this conspiracy theory!")
    lcd.initiate_message(100, "COVID-19 vaccine contains microchips. This is a hoax!")

    target_node = int(input("\nEnter a target node: "))
    message_content = input("Enter a message for the target node (press Enter for default): ")

    lcd.analyze_message_impact(target_node, message_content)

    node_info = lcd.get_node_info(target_node)

    print(f"\nTarget Node: {target_node}")
    print(f"Community: {node_info.community}")
    print(f"Number of connected communities: {len(node_info.connected_communities)}")
    print(f"Number of directly connected nodes: {len(node_info.directly_connected_nodes)}")
    print(f"Total number of connected nodes from all communities: {len(node_info.all_connected_nodes)}")

    print("Directly connected nodes:", end=" ")
    print(*node_info.directly_connected_nodes)

if __name__ == "__main__":
    main()